dbt/adapters/bigquery/relation_configs/_options.py

```diff
@@ -158,4 +176,19 @@ class BigQueryOptionsConfigChange(RelationConfigChange):
 
     @property
     def requires_full_refresh(self) -> bool:
-        return False
+        # allow_non_incremental_definition cannot be changed via an ALTER statement
+        return self.context.allow_non_incremental_definition is not None
+
+    @classmethod
+    def from_options_configs(
+        cls, new_options: BigQueryOptionsConfig, existing_options: BigQueryOptionsConfig
+    ) -> Self:
+        new_options_dict = asdict(new_options)
+        existing_options_dict = asdict(existing_options)
+        option_diffs_dict = {
+            k: v
+            for k, v in new_options_dict.items()
+            if new_options_dict[k] != existing_options_dict[k]
+        }
+        option_diffs = BigQueryOptionsConfig.from_dict(option_diffs_dict)
+        return cls(action=RelationConfigChangeAction.alter, context=option_diffs)
```

Comment:
Oh I think what you were doing ahead now. I'm not sure I like burrying the alter here. May be a matter of preference. I kind of what to float this action param up for visibility when creating the object as opposed to hardcoding the value down here.

Reply:
I don't understand, but I'll comment with what's coming to mind.

Most things are an `alter` in this framework. The other options are `create` and `drop`, which really was only used for indexes in `dbt-postgres` where the entire object (index, not the relation) would be created or dropped instead of altered. To alter an index in `dbt-postgres` you need to issue two statements, one to drop the existing, and another to recreate it with the new config. In this case, we're really just altering the options on the relation (in this case an MV, but could be a table in the future). Sometimes that happens via a full refresh, which gets determined based on the context.

I agree we could spend more time on this framework to make it easier to parse (and I would like to do so if given capacity), but updating that would be outside of the scope of this particular change. Hence I left it with the default `alter` since I need to provide something. Unfortunately, if you follow the code through, this attribute never gets used, only `context` does.

Reply:
I agree on not scoping up. I'd just rather have an empty param here and pass in the `alter` deliberately at the function call level above. Right now you're making an assumption that I worry may not be obvious to the next person. But with the generic param passing this argument from the function call, it would be quite straightforward for adding others as needed (even if it never happens, I think it'd read clearer)

Reply:
When you say to pass it in to the function call above, are you saying to pass it in to `from_options_configs`? While it's static now, it may not be in the future. In fact, I could see accounting for "unsetting" options in the future, would would just be a drop. Then I would need to compare the two options configs to see if it's an alter or a drop/create scenario. That would make `from_options_configs` duplicative. My intent for this method was basically "here are my two options configs, please tell me what I need to do to get from A to B". I don't think supplying the action as an argument as well makes it more readable. I could set the action prior to the return with a comment that all diffs are alter diffs at this point, which may change in the future.

Reply:
With the change of providing the entire options config as context for the change, there is no need for a `from_options_config` method. Hence this is now back in the calling method. I think this is what you were asking for, but please let me know if that's incorrect.